### ISSUE TITLE
Job stays pending for too long

### ISSUE THREAD

**Issue by mildred** [hashicorp/nomad#2153]
If you have a question, prepend your issue with `[question]` or preferably use the [nomad mailing list](https://www.nomadproject.io/community.html).

If filing a bug please include the following:

- Nomad version: v0.4.1
- Operating system and Environment details: CoreOS

### Issue

Sometimes, a job takes too much time in pending state. Sometimes it immediatly starts, other times it can take more then 10 minutes, and I have memories of few hours pending.

This is not necessarily a problem, if that delay can be explained and be reduced (for example if it comes from downloading the docker image). Unfortunately, there is not much information on what is happening in pending state and one is left wondering if the job will start at all.

To help debug this, what is happening during the pending phase ? I suppose that there is the placement, shouldn't take that much time. There is also probably the docker image download (that should be the same each time I suppose). Does nomad wait for anything else before running the job ? Does it wait for the nodes to have enough free resources or not ? Can it block ?

Here is the alloc-status of a job that took 5 minutes to start:

```
ID            = bf237918
Eval ID       = d3d2c5d2
Name          = fae38a702a03ad57b4f4b6783b13b3fea03e962a:[staging][shanti-dev3][mildred/rails-demo-1].sqsc-job[0]
Node ID       = 86ab32e7
Job ID        = fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand
Client Status = complete

Task "fae38a702a03ad57b4f4b6783b13b3fea03e962a-task-precommand" is "dead"
Task Resources
CPU      Memory   Disk     IOPS  Addresses
500 MHz  256 MiB  300 MiB  0     

Recent Events:
Time                   Type        Description
01/03/17 09:47:42 UTC  Terminated  Exit Code: 0
01/03/17 09:46:30 UTC  Started     Task started by client
01/03/17 09:41:52 UTC  Received    Task received by client
```

### Reproduction steps

Difficult to reproduce, but the jobs were submitted using the HTTP API.

### Nomad Server logs (if appropriate)

Running many nomad clients with a server cluster of 3 nodes. I don't have the full logs for all (there was a reboot) but for the 2nd node, I have:

```
Jan 03 09:40:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:40:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.51:4648
Jan 03 09:41:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:58842
Jan 03 09:41:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:41:52 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:52.426276 [DEBUG] worker: dequeued evaluation edf9717a-97ac-dc84-ced2-296217fe07a4
Jan 03 09:41:52 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:52.526686 [DEBUG] sched: <Eval 'edf9717a-97ac-dc84-ced2-296217fe07a4' JobID: 'fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>: allocs: (place 0) (update 0) (migrate 0) (stop 0) (ignore 0) (lost 0)
Jan 03 09:41:52 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:52.526708 [DEBUG] sched: <Eval 'edf9717a-97ac-dc84-ced2-296217fe07a4' JobID: 'fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>: setting status to complete
Jan 03 09:41:52 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:52.528634 [DEBUG] worker: updated evaluation <Eval 'edf9717a-97ac-dc84-ced2-296217fe07a4' JobID: 'fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>
Jan 03 09:41:52 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:41:52.529415 [DEBUG] worker: ack for evaluation edf9717a-97ac-dc84-ced2-296217fe07a4
Jan 03 09:42:32 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:42:32 [DEBUG] memberlist: TCP connection from=10.0.0.51:38126
Jan 03 09:42:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:42:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:58920
Jan 03 09:42:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:42:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.51:4648
Jan 03 09:43:18 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:18.807987 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (8.406555ms)
Jan 03 09:43:18 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:18.810688 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.237695ms)
Jan 03 09:43:18 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:18.819665 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (7.67815ms)
Jan 03 09:43:32 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:32 [DEBUG] memberlist: TCP connection from=10.0.0.51:38220
Jan 03 09:43:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:59016
Jan 03 09:43:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:43:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:44:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:44:43.912147 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.310183ms)
Jan 03 09:44:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:44:43.914354 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.11247ms)
Jan 03 09:44:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:44:43.916657 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (1.003206ms)
Jan 03 09:44:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:44:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:45:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:45:50.384958 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (8.188541ms)
Jan 03 09:45:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:45:50.387868 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.345695ms)
Jan 03 09:45:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:45:50.390217 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (1.008558ms)
Jan 03 09:45:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:45:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:46:12 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:12.004093 [DEBUG] worker: dequeued evaluation ec6cb7b9-db56-d590-b8a4-efda0132630f
Jan 03 09:46:12 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:12.004582 [DEBUG] sched.core: job GC: scanning before index 224918 (4h0m0s)
Jan 03 09:46:12 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:12.005482 [DEBUG] worker: ack for evaluation ec6cb7b9-db56-d590-b8a4-efda0132630f
Jan 03 09:46:32 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:32 [DEBUG] memberlist: TCP connection from=10.0.0.51:38492
Jan 03 09:46:43 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:59262
Jan 03 09:46:50 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:50 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:46:58 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:58.425537 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.264033ms)
Jan 03 09:46:58 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:58.427837 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.142985ms)
Jan 03 09:46:58 ip-10-0-0-36.eu-west-1.compute.internal nomad[1070]:     2017/01/03 09:46:58.430145 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (968.629µs)
```

and the logs for the server node #3:

```
Jan 03 09:40:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:40:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:40:43 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:40:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:52968
Jan 03 09:40:50 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:40:50 [DEBUG] memberlist: TCP connection from=10.0.0.36:38338
Jan 03 09:41:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.676630 [DEBUG] worker: dequeued evaluation d3d2c5d2-2df4-44c4-7a04-e411ad2264
9a
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.777559 [DEBUG] sched: <Eval 'd3d2c5d2-2df4-44c4-7a04-e411ad22649a' JobID: 'fa
e38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>: allocs: (place 1) (update 0) (migrate 0) (stop 0) (ignore 0) (lost 0)
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.780797 [DEBUG] worker: submitted plan for evaluation d3d2c5d2-2df4-44c4-7a04-
e411ad22649a
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.781085 [DEBUG] sched: <Eval 'd3d2c5d2-2df4-44c4-7a04-e411ad22649a' JobID: 'fa
e38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>: setting status to complete
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.783008 [DEBUG] worker: updated evaluation <Eval 'd3d2c5d2-2df4-44c4-7a04-e411
ad22649a' JobID: 'fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand'>
Jan 03 09:41:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:41:52.783965 [DEBUG] worker: ack for evaluation d3d2c5d2-2df4-44c4-7a04-e411ad22649
a
Jan 03 09:42:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:42:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.36:4648
Jan 03 09:42:50 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:42:50 [DEBUG] memberlist: TCP connection from=10.0.0.36:38496
Jan 03 09:43:09 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:43:09.761212 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.390601ms)
Jan 03 09:43:09 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:43:09.764375 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a4
4f7 (1.105496ms)
Jan 03 09:43:09 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:43:09.766902 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (1
.247369ms)
Jan 03 09:43:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:43:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.36:4648
Jan 03 09:44:21 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:44:21.709016 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.107084ms)
Jan 03 09:44:21 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:44:21.711206 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a4
4f7 (1.079572ms)
Jan 03 09:44:21 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:44:21.713499 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (8
81.468µs)
Jan 03 09:44:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:44:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:44:43 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:44:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:53294
Jan 03 09:45:15 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:45:15.314983 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.369619ms)
Jan 03 09:45:15 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:45:15.317270 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.13353ms)
Jan 03 09:45:15 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:45:15.319571 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (1.018653ms)
Jan 03 09:45:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:45:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.20:4648
Jan 03 09:45:43 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:45:43 [DEBUG] memberlist: TCP connection from=10.0.0.20:53380
Jan 03 09:46:12 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:12.004415 [DEBUG] worker: dequeued evaluation 9ed35f34-ca15-9549-fcf6-e1eae2dc9aa2
Jan 03 09:46:12 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:12.004938 [DEBUG] sched.core: eval GC: scanning before index 224918 (1h0m0s)
Jan 03 09:46:12 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:12.006063 [DEBUG] worker: ack for evaluation 9ed35f34-ca15-9549-fcf6-e1eae2dc9aa2
Jan 03 09:46:32 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:32 [DEBUG] memberlist: Initiating push/pull sync with: 10.0.0.36:4648
Jan 03 09:46:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:52.327024 [DEBUG] http: Request /v1/allocations?prefix=bf237918 (1.747868ms)
Jan 03 09:46:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:52.329450 [DEBUG] http: Request /v1/allocation/bf237918-6d8e-15e4-27a7-54aa568a44f7 (1.249612ms)
Jan 03 09:46:52 ip-10-0-0-51.eu-west-1.compute.internal nomad[1092]:     2017/01/03 09:46:52.331718 [DEBUG] http: Request /v1/node/86ab32e7-8bda-8d55-25d4-c23837ddc46b (942.445µs)
```

### Nomad Client logs (if appropriate)

Unfortunately I no longer have access to the nomad client logs. The machine was stopped. I remember I looked at it and it was quite empty. So much so that I looked at the docker daemon logs. When I have this problem again, I'll make sure to include the logs.

### Job file (if appropriate)

I don't have the exact job file, but it looks like this one :

```
{
   "Job" : {
      "Region" : "europe",
      "AllAtOnce" : false,
      "JobModifyIndex" : 0,
      "Constraints" : [
         {
            "Operand" : "=",
            "LTarget" : "${meta.sqsc.environment}",
            "RTarget" : "staging"
         },
         {
            "RTarget" : "shanti-dev3",
            "Operand" : "=",
            "LTarget" : "${meta.sqsc.project}"
         },
         {
            "LTarget" : "${attr.kernel.name}",
            "Operand" : "=",
            "RTarget" : "linux"
         }
      ],
      "Datacenters" : [
         "staging-sqsc"
      ],
      "ModifyIndex" : 0,
      "Type" : "batch",
      "TaskGroups" : [
         {
            "Tasks" : [
               {
                  "Name" : "fae38a702a03ad57b4f4b6783b13b3fea03e962a-task-precommand",
                  "LogConfig" : {
                     "MaxFileSizeMB" : 10,
                     "MaxFiles" : 10
                  },
                  "Meta" : null,
                  "Resources" : {
                     "CPU" : 500,
                     "MemoryMB" : 256,
                     "IOPS" : 0,
                     "Networks" : [
                        {
                           "MBits" : 10,
                           "Public" : false,
                           "CIDR" : "",
                           "IP" : "",
                           "ReservedPorts" : [],
                           "DynamicPorts" : null
                        }
                     ],
                     "DiskMB" : 300
                  },
                  "Env" : {},
                  "Artifacts" : null,
                  "Constraints" : null,
                  "Config" : {
                     "image" : "095348363195.dkr.ecr.eu-west-1.amazonaws.com/client-shanti-dev3-staging-mildred-rails-demo-1:v0",
                     "command" : "bundle",
                     "args" : [
                        "exec",
                        "rails",
                        "db:migrate"
                     ],
                     "port_map" : []
                  },
                  "KillTimeout" : 5000000000,
                  "Services" : [],
                  "Driver" : "docker",
                  "User" : ""
               }
            ],
            "Constraints" : null,
            "Count" : 1,
            "RestartPolicy" : {
               "Delay" : 1000000000,
               "Attempts" : 0,
               "Mode" : "fail",
               "Interval" : 1000000000
            },
            "Meta" : null,
            "Name" : "sqsc-job"
         }
      ],
      "CreateIndex" : 0,
      "StatusDescription" : "",
      "Update" : {
         "Stagger" : 10000000000,
         "MaxParallel" : 1
      },
      "Periodic" : null,
      "Priority" : 75,
      "Meta" : null,
      "Status" : "",
      "Name" : "fae38a702a03ad57b4f4b6783b13b3fea03e962a:[staging][shanti-dev3][mildred/rails-demo-1]",
      "ID" : "fae38a702a03ad57b4f4b6783b13b3fea03e962a-precommand"
   }
}

```

**Comment by dadgar** [comment]
Hey @mildred,

How large is that image? It is most likely downloading the image. In 0.5.3 we will have drivers emit extra information to make debugging this easier (it will show up in alloc-status events). 

The reason I believe it to be downloading the image is that once the client marks an allocation as `received` it immediately invokes the driver and the first thing the docker driver does is download/check for the image and once that is done starts it. When it starts that is the second event.

**Comment by dadgar** [comment]
I am not sure what you would like to do with this issue. We can maybe close it till you run into it again and post the logs (I would suggest running in debug level logs) so we can be more certain that it is the downloading the image or wait for 0.5.3 which should be out by end of the month.

**Comment by dadgar** [comment]
Closed by https://github.com/hashicorp/nomad/pull/2177

**Comment by maximveksler** [comment]
I'd suggest a feature request where docker would first download the image, and only then nomad would kill existing instances (in case of running only a single instance). 

Otherwise for large images there is a potential downtime while the image is being downloaded.

**Comment by schmichael** [comment]
@maximveksler If you're referring to downtime caused by job updates stopping old allocations before starting new ones the [`update` stanza](https://www.nomadproject.io/docs/job-specification/update.html) and a `count > 1` is the suggested solution for ensuring services are available during updates. Please feel free to open a new issue, [post to the mailing list](https://groups.google.com/group/nomad-tool), or ask in [gitter](https://gitter.im/hashicorp-nomad/Lobby) if you still have concerns.
